### PR TITLE
handle multiple changes in one watch event

### LINKED ISSUE #2581: e2e - "should surface components in its status" failure
## Bug Report

e2e test "should surface components in its status" failed
```
Operator API 
  should surface components in its status
  /home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/operator_test.go:73

[BeforeEach] Operator API
  /home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/operator_test.go:41
[It] should surface components in its status
  /home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/operator_test.go:73
STEP: eventually having a status that contains its component label selector
STEP: eventually listing a single component reference
STEP: eventually listing multiple component references


• Failure [90.469 seconds]
Operator API
/home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/operator_test.go:31
  should surface components in its status [It]
  /home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/operator_test.go:73

  Unexpected error:
      <context.deadlineExceededError>: {}
      context deadline exceeded
  occurred

  /home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/util.go:892

  Full Stack Trace
  github.com/operator-framework/operator-lifecycle-manager/test/e2e.awaitPredicates({0x32ac0b0, 0xc001e0fec0}, {0x328d2d8, 0xc001f8a200}, {0xc00114cc78, 0x1, 0x328ba28})
  	/home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/util.go:892 +0x213
  github.com/operator-framework/operator-lifecycle-manager/test/e2e.componentRefEventuallyExists({0x328d2d8, 0xc001f8a200}, 0x1, 0xc001da96c0)
  	/home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/operator_test.go:408 +0xfc
  github.com/operator-framework/operator-lifecycle-manager/test/e2e.glob..func19.2()
  	/home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/operator_test.go:181 +0xe93
  github.com/operator-framework/operator-lifecycle-manager/test/e2e.TestEndToEnd(0xf908f9)
  	/home/runner/work/operator-lifecycle-manager/operator-lifecycle-manager/test/e2e/e2e_test.go:59 +0x268
  testing.tRunner(0xc000aac4e0, 0x3004308)
  	/opt/hostedtoolcache/go/1.17.6/x64/src/testing/testing.go:1259 +0x102
  created by testing.(*T).Run
  	/opt/hostedtoolcache/go/1.17.6/x64/src/testing/testing.go:1306 +0x35a
```

**What did you do?**
Looked at CI e2e test failure

**What did you expect to see?**
No failure

**What did you see instead? Under which circumstances?**
Error above

**Environment**
* operator-lifecycle-manager version:

* Kubernetes version information:

* Kubernetes cluster kind: 

**Possible Solution**

**Additional context**

### PR DESCRIPTION
**Pull request by akihikokuroda**
Signed-off-by: akihikokuroda <akihikokuroda2020@gmail.com>

**Description of the change:**

When waiting for the multiple components added in the status, the changes for the operator instance sometimes come in one watch event.  The original test could not handle this case and wait for following watch event that may not come. 

**Motivation for the change:**
Closes #2581 
duplicate issue: https://github.com/operator-framework/operator-lifecycle-manager/pull/2665
**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/doc`
- [ ] Commit messages sensible and descriptive
- [ ] Tests marked as `[FLAKE]` are truly flaky
- [ ] Tests that remove the `[FLAKE]` tag are no longer flaky

### CODE REVIEW

File: test/e2e/operator_test.go [hunk 1]
```diff
@@ -421,6 +420,31 @@ func componentRefEventuallyExists(w watch.Interface, exists bool, ref *corev1.Ob
...
+				if !founds[i] && r.APIVersion == refs[i].APIVersion && r.Kind == refs[i].Kind && r.Namespace == refs[i].Namespace && r.Name == refs[i].Name {
+					founds[i] = true
+				}
+			}
+		}
+		for _, found := range founds {
+			if !found {
+				return !exists
+			}
+		}
```
Comment (awgreene) [inline test/e2e/operator_test.go:443]: Nit: I found this a bit difficult to read.

At the cost of memory, we might be able to make this code more intuitive by creating a `map[corev1.ObjectReference]struct{}` that tracks each unseen ref and removing each key as it is seen. Once the length of the map is zero, we know that we've seen all elements.

Comment (akihikokuroda) [reply test/e2e/operator_test.go]: Thanks for review.  I updated the code.  I hope this is more readable than the previous one.

File: test/e2e/operator_test.go [hunk 2]
```diff
@@ -421,6 +420,30 @@ func componentRefEventuallyExists(w watch.Interface, exists bool, ref *corev1.Ob
...
+					refs[i] = nil
+				}
+			}
+		}
+		for _, ref := range refs {
+			if ref != nil {
+				return !exists
+			}
+		}
+		return exists
```
Comment (perdasilva) [inline test/e2e/operator_test.go:443]: I think mutating the status could be dangerous. How about something like this?

```suggestion
        foundCount := 0
	awaitPredicates(deadline, w, operatorPredicate(func(op *operatorsv1.Operator) bool {
		if op.Status.Components == nil {
			return false
		}
		for _, r := range op.Status.Components.Refs {
			for i := 0; i < len(refs); i++ {
				if refs[i] != nil && r.APIVersion == refs[i].APIVersion && r.Kind == refs[i].Kind && r.Namespace == refs[i].Namespace && r.Name == refs[i].Name {
					foundCount++
				}
			}
		}
                return foundCount == len(refs)
```

File: test/e2e/operator_test.go [hunk 3]
```diff
@@ -421,6 +420,30 @@ func componentRefEventuallyExists(w watch.Interface, exists bool, ref *corev1.Ob
 	}))
 }
 
+func multipleComponentRefsEventuallyExists(w watch.Interface, exists bool, refs []*corev1.ObjectReference) {
+	deadline, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancel()
```
Comment (timflannagan) [inline test/e2e/operator_test.go:425]: Hmm let the test case handle this instead, and take a context.Context parameter? Thoughts?

File: test/e2e/operator_test.go [hunk 4]
```diff
@@ -421,6 +420,30 @@ func componentRefEventuallyExists(w watch.Interface, exists bool, ref *corev1.Ob
 	}))
 }
 
+func multipleComponentRefsEventuallyExists(w watch.Interface, exists bool, refs []*corev1.ObjectReference) {
```
Comment (timflannagan) [inline test/e2e/operator_test.go:423]: Update refs to be a variadic parameter?

File: test/e2e/operator_test.go (hunk 3 above)
Comment (akihikokuroda) [reply test/e2e/operator_test.go:425]: I don't have a strong opinion about this.  It's the same way as the `componentRefEventuallyExists` function.  I can refactor it, too.

File: test/e2e/operator_test.go (hunk 4 above)
Comment (akihikokuroda) [reply test/e2e/operator_test.go:423]: I'll update.  Thanks!

File: test/e2e/operator_test.go [hunk 5]
```diff
@@ -421,6 +425,27 @@ func componentRefEventuallyExists(w watch.Interface, exists bool, ref *corev1.Ob
...
+				}
+			}
+		}
+		for _, ref := range refs {
+			if ref != nil {
+				return !exists
+			}
+		}
+		return exists
+	}))
```
Comment (njhale) [inline test/e2e/operator_test.go:446]:
```suggestion
	awaitPredicates(ctx, w, operatorPredicate(func(op *operatorsv1.Operator) bool {
		if op.Status.Components == nil {
			return false
		}
		
		var found int
		// Note: This is roughly O(n*m) but is good enough for testing purposes.
		for _, expected := range refs {
		    for _, actual := range op.Status.Components.Refs {
		        if actual.APIVersion == expected.APIVersion && actual.Kind == expected.Kind && actual.Namespace == expected.Namespace && actual.Name == expected.Name {
			    found++
			    break // Found it, look for next expected	
                        }
		    }
		}
		
		if exists {
		    return found == len(refs)
		}
		
		// Expect nothing has been found
		return found == 0
	}))
```

File: test/e2e/operator_test.go [hunk 6]
```diff
@@ -147,7 +147,9 @@ var _ = Describe("Operator API", func() {
 
 		// Ensure o's status.components.refs field eventually contains a reference to ns-a
 		By("eventually listing a single component reference")
-		componentRefEventuallyExists(w, true, getReference(scheme, nsA))
+		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
```
Comment (njhale) [inline test/e2e/operator_test.go:150]: I think this should be using `clientCtx` instead of `context.Background()`.

Comment (njhale) [inline test/e2e/operator_test.go:150]: `deadline` is the convention for timeout contexts -- instead of `ctx` -- elsewhere in this suite.

**Review by njhale — CHANGES_REQUESTED** [review]
Nice work refactoring! I have a few small requests, but looking good so far.

File: test/e2e/operator_test.go [hunk 7]
```diff
@@ -178,8 +180,9 @@ var _ = Describe("Operator API", func() {
 
 		// Ensure o's status.components.refs field eventually contains references to sa-a and sa-b
 		By("eventually listing multiple component references")
-		componentRefEventuallyExists(w, true, getReference(scheme, saA))
-		componentRefEventuallyExists(w, true, getReference(scheme, saB))
+		ctx, cancel = context.WithTimeout(context.Background(), 1*time.Minute)
+		multipleComponentRefsEventuallyExists(ctx, w, true, []*corev1.ObjectReference{getReference(scheme, saA), getReference(scheme, saB)}...)
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply test/e2e/operator_test.go:184]: Thanks!  I'll change it.

File: test/e2e/operator_test.go (hunk 5 above)
Comment (akihikokuroda) [reply test/e2e/operator_test.go:446]: Thanks!  I take this suggestion.

File: test/e2e/operator_test.go (hunk 6 above)
Comment (akihikokuroda) [reply test/e2e/operator_test.go:150]: I change them.  Thanks!

Comment (akihikokuroda) [reply test/e2e/operator_test.go:150]: I change them.  Thanks!

**Comment by perdasilva** [comment]
@akihikokuroda is this ready for re-review? FYI, I've rebased from this side ^^

**Comment by perdasilva** [comment]
closing PR as stale. Please re-open if it's still important.

**Pull request closed without merging**